### ISSUE TITLE
prepared query returns empty list when combining tag and connect filter

### ISSUE THREAD

**Issue by ekim38** [hashicorp/consul#4700]
Consul connect is great, and i've been playing around with it to implement a singleton service. my strategy was to elect a leader, tag that service, which then would be selected by prepared query using [this](https://www.consul.io/docs/connect/proxies.html#prepared-query-upstreams).

#### Overview of the Issue

The issue is prepared query won't return the tagged service. I saw in log that connect proxy is making query api call with connect=true parameter, so I reproduced it by making that same api call. I've ran 4 variations and saw that it didn't work only when I combined tag and connect filters together.

1. no tag + no connect filter: returns the service [log](https://gist.github.com/ekim38/c1a9aec00f98d193a4f8d0dedb4f2c99#file-no_tag__no_connect__query)
2. no tag + connect filter: returns the service [log](https://gist.github.com/ekim38/c1a9aec00f98d193a4f8d0dedb4f2c99#file-no_tag__connect__query) <- this tells me the service is indeed connect capable
3. tag + no connect filter: returns the service [log](https://gist.github.com/ekim38/c1a9aec00f98d193a4f8d0dedb4f2c99#file-tag__no_connect__query) <- this tells me the service is tagged appropriately
4. tag + connect filter: doesn't return the service [log](https://gist.github.com/ekim38/c1a9aec00f98d193a4f8d0dedb4f2c99#file-tag__connect__query)

#### Reproduction Steps

Steps to reproduce this issue, eg:

1. vanilla dev cluster (v 1.2.3)
2. vanilla service with connect proxy and a tag [log](https://gist.github.com/ekim38/c1a9aec00f98d193a4f8d0dedb4f2c99#file-services-json)
3. two prepared queries. one has no tag filter, one has tag filter [log](https://gist.github.com/ekim38/c1a9aec00f98d193a4f8d0dedb4f2c99#file-queries-json)
4. execute each query with and without connect=true parameter

### Consul info for both Client and Server

<details>
  <summary>Client info</summary>

agent:
	check_monitors = 0
	check_ttls = 0
	checks = 1
	services = 2
build:
	prerelease =
	revision = 48d287ef
	version = 1.2.3
consul:
	known_servers = 1
	server = false
runtime:
	arch = amd64
	cpu_count = 1
	goroutines = 56
	max_procs = 1
	os = linux
	version = go1.10.1
serf_lan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 2
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 2
	members = 2
	query_queue = 0
	query_time = 1

</details>

<details>
  <summary>Server info</summary>

agent:
	check_monitors = 0
	check_ttls = 0
	checks = 0
	services = 0
build:
	prerelease =
	revision = 48d287ef
	version = 1.2.3
consul:
	bootstrap = true
	known_datacenters = 1
	leader = true
	leader_addr = 192.168.33.10:8300
	server = true
raft:
	applied_index = 429
	commit_index = 429
	fsm_pending = 0
	last_contact = 0
	last_log_index = 429
	last_log_term = 2
	last_snapshot_index = 0
	last_snapshot_term = 0
	latest_configuration = [{Suffrage:Voter ID:46324d7a-101e-dd07-ccc4-0328ef311264 Address:192.168.33.10:8300}]
	latest_configuration_index = 1
	num_peers = 0
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Leader
	term = 2
runtime:
	arch = amd64
	cpu_count = 1
	goroutines = 82
	max_procs = 1
	os = linux
	version = go1.10.1
serf_lan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 2
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 2
	members = 2
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 1
	members = 1
	query_queue = 0
	query_time = 1

</details>

### Operating system and Environment details

Vagrant + virtualbox + centos/7 + consul linux 64bit v1.2.3

**Comment by banks** [comment]
Thanks for the detailed report and repro case!

I spent a while looking before I realised that this is _kinda_ intended behaviour but something we do a bad job of documenting (and probably the wrong default UX).

The issue is that you didn't tag the _proxy service_ only the actual destination. The way `Connect: true` prepared queries work is actually just by looking at all service registrations that support connect (either proxies, or native integrations) and whose _target service_ is the requested service name.

The corollary is that all other metadata like tag or service meta matches are made against the _proxy_ service instance not the target one.

This is partly a result of the internal model for the data, but there could be an argument that this gives most control - it allows a service that is migrating to connect but still has non-connect clients to separately control tags for the connect vs non-connect clients.

But it's also not ideal UX in terms of least surprise and even I spent a while looking before I realised what was going on!

We'll consider this more closely. It's not clear if we just need to document the model of what's happening here better, or change how it actually works. Although it seems trivial, it's actually an important part of the internal model that there is not necessarily a way to do the "normal" service lookup and then work backwards to a matching connect proxy since there may be many proxies, some sidecars some central loadbalancers, that eventually resolve to a given instance.

We could potentially make a special case for tags or default to copying tags into proxy definitions but both of those would be somewhat magical and "surprising" in a different way.

**tl;dr**: for now add the `foo` tag to `web-proxy` service too and all will work. We will consider how to make this more obvious.
